test/e2e_node/gci-init.sh

```diff
@@ -16,36 +16,48 @@
 
 # This script is used to do extra initialization on GCI.
 
-# set up cgroupv2 based on flag CONTAINERD_CGROUPV2 in containerd-env
 CONTAINERD_HOME="/home/containerd"
 CONTAINERD_ENV_METADATA="containerd-env"
 
 if [ -f "${CONTAINERD_HOME}/${CONTAINERD_ENV_METADATA}" ]; then
   source "${CONTAINERD_HOME}/${CONTAINERD_ENV_METADATA}"
 fi
 
-if [ "${CONTAINERD_CGROUPV2:-"false"}"  == "true" ]; then
-  # check cos image
-  if [ -r /etc/os-release ]; then
-    OS_ID="$(. /etc/os-release && echo "$ID")"
+function configure_cgroup_mode() {
+  if [[ -z "${CONTAINERD_COS_CGROUP_MODE}" ]]; then
+    return
   fi
-  if [ "${OS_ID}" = "cos" ]; then
-    if ! grep -q 'systemd.unified_cgroup_hierarchy=true' /proc/cmdline && [ "$(stat -fc %T /sys/fs/cgroup/)" != "cgroup2fs" ]; then
-      echo "Setting up cgroupv2"
 
-      mount_path="/tmp/esp"
-      mkdir -p "${mount_path}"
-      esp_partition="/dev/sda12"
-      mount "${esp_partition}" "${mount_path}"
-      sed -i 's/systemd.unified_cgroup_hierarchy=false/systemd.unified_cgroup_hierarchy=true/g' "${mount_path}/efi/boot/grub.cfg"
-      umount "${mount_path}"
-      rmdir "${mount_path}"
+  if [[ ! -r /etc/os-release ]]; then
+    return
+  fi
 
-      echo "Reconfigured grub; rebooting..."
-      reboot
-    fi
+  OS_ID="$(. /etc/os-release && echo "$ID")"
+  if [[ "${OS_ID}" != "cos" ]]; then
+    return
   fi
-fi
+
+  # cgroup_helper was introduced in COS M97, see if it's available first...
+  if ! command -v cgroup_helper > /dev/null 2>&1; then
+    return
+  fi
+
+  # if cgroup mode requested was v1 but it's currently set as unified (v2),
+  # switch to hybrid (v1) and reboot
+  if [[ "${CONTAINERD_COS_CGROUP_MODE:-}" == "v1" ]] && cgroup_helper show | grep -q 'unified'; then
+    cgroup_helper set hybrid
+    echo "set cgroup config to hybrid, now rebooting..."
+    reboot
```

Comment:
Since you mentioned that this script is invoked via cloud-init: is cloud-init configured to execute it on each boot rather than just the first boot?  I know most distributions' cloud-init would default to executing a script only on the first boot on EC2, but I don't know how this is configured on COS or GCE.

If the script doesn't get executed after the reboot, lines 62-78 won't get executed in the case where the cgroup mode needs to be reconfigured.

Reply:
Yes, COS is configured to run the `runcmd` and `write_files` section of a cloud-config on each boot. xref: https://cloud.google.com/container-optimized-os/docs/how-to/create-configure-instance

> Note: Using systemctl enable in the runcmd section of cloud-init will not work as intended. This is because /etc on Container-Optimized OS is stateless, so commands such as systemctl enable/disable, that modify /etc, won't persist across reboots. See [Disks and Filesystem](https://cloud.google.com/container-optimized-os/docs/concepts/disks-and-filesystem) for more information on Container-Optimized OS's filesystem. An alternative to systemctl enable is to use systemctl start, as in the example above. **cloud-init modules such as write_files and runcmd, which are typically run once-per-instance on other distros, are run on every boot on Container-Optimized OS.**

Reply:
Thanks for confirming!